cartographer_grpc/framework/rpc.cc

```diff
@@ -101,29 +128,95 @@ void Rpc::RequestStreamingReadIfNeeded() {
 }
 
 void Rpc::Write(std::unique_ptr<::google::protobuf::Message> message) {
-  response_ = std::move(message);
   switch (rpc_handler_info_.rpc_type) {
+    case ::grpc::internal::RpcMethod::BIDI_STREAMING:
+      // For BIDI_STREAMING enqueue the message into the send queue and
+      // start write operations if none are currently in-flight.
+      send_queue_.emplace(SendItem{std::move(message), ::grpc::Status::OK});
+      PerformWriteIfNeeded();
+      break;
+    case ::grpc::internal::RpcMethod::CLIENT_STREAMING:
+      SendFinish(::grpc::Status::OK, std::move(message));
+      break;
+    case ::grpc::internal::RpcMethod::NORMAL_RPC:
+      SendFinish(::grpc::Status::OK, std::move(message));
+      break;
+    default:
+      LOG(FATAL) << "RPC type not implemented.";
+  }
+}
+
+void Rpc::SendFinish(::grpc::Status status,
+                     std::unique_ptr<::google::protobuf::Message> message) {
+  switch (rpc_handler_info_.rpc_type) {
+    case ::grpc::internal::RpcMethod::BIDI_STREAMING:
+      CHECK(!message);
+      finish_event_.pending = true;
```

Comment:
If this is happens in all cases, consider moving before switch.

Reply:
Done.